base/intfuncs.jl

```diff
@@ -195,29 +195,28 @@ end
 ^(x::Number, p::Integer)  = power_by_squaring(x,p)
 ^(x, p::Integer)          = power_by_squaring(x,p)
 
-# x^p for any literal integer p is lowered to x^Val{p},
+# x^p for any literal integer p is lowered to Base.literal_pow(^, x, Val{p})
 # to enable compile-time optimizations specialized to p.
-# However, we still need a fallback that calls the general ^.
-# To avoid ambiguities for methods that dispatch on the
-# first argument, we dispatch the fallback via internal_pow.
+# However, we still need a fallback that calls the function ^ which may either
+# mean Base.^ or something else, depending on context.
 # We mark these @inline since if the target is marked @inline,
 # we want to make sure that gets propagated,
 # even if it is over the inlining threshold.
-@inline ^(x, p) = internal_pow(x, p)
-@inline internal_pow{p}(x, ::Type{Val{p}}) = x^p
+@inline literal_pow{p}(f, x, ::Type{Val{p}}) = f(x,p)
 
 # Restrict inlining to hardware-supported arithmetic types, which
-# are fast enough to benefit from inlining.  This also makes it
-# easier to override ^ without having to override the Val method.
+# are fast enough to benefit from inlining.
 const HWReal = Union{Int8,Int16,Int32,Int64,UInt8,UInt16,UInt32,UInt64,Float32,Float64}
 const HWNumber = Union{HWReal, Complex{<:HWReal}, Rational{<:HWReal}}
 
 # inference.jl has complicated logic to inline x^2 and x^3 for
-# numeric types.  In terms of Val we can do it much more simply:
-@inline internal_pow(x::HWNumber, ::Type{Val{0}}) = one(x)
-@inline internal_pow(x::HWNumber, ::Type{Val{1}}) = x
-@inline internal_pow(x::HWNumber, ::Type{Val{2}}) = x*x
-@inline internal_pow(x::HWNumber, ::Type{Val{3}}) = x*x*x
+# numeric types.  In terms of Val we can do it much more simply.
+# (The first argument prevents unexpected behavior if a function ^
+# is defined that is not equal to Base.^)
+@inline literal_pow(::typeof(^), x::HWNumber, ::Type{Val{0}}) = one(x)
+@inline literal_pow(::typeof(^), x::HWNumber, ::Type{Val{1}}) = x
+@inline literal_pow(::typeof(^), x::HWNumber, ::Type{Val{2}}) = x*x
+@inline literal_pow(::typeof(^), x::HWNumber, ::Type{Val{3}}) = x*x*x
 
```

Comment:
Is there a particular reason to stop at 3 rather than e.g. 4? Of course there's virtually no limit, but 4 seems like a common power (much more than 5 and above I'd say). This could help with things like https://github.com/JuliaStats/HypothesisTests.jl/pull/238.

Reply:
There's an increasing loss of accuracy at higher powers. This definition for 3 is already inaccurate by a noticeable bit in some cases, so 4+ will potentially be progressively worse from there (though `(x*x)^2` seems pretty well-compensated due to the symmetric shape, as these things go).

Reply:
OK, thanks. So are you suggesting we should add a definition using `(x*x)^2`?